transformer_engine/jax/cpp_extensions/attention.py

```diff
@@ -1020,6 +895,361 @@ def sharded_impl(
+
+                kv_seqlens_for_rank = helper.kv_seqlens_for_rank(
+                    idx, kv_max_seqlen, kv_seqlen_per_subrank
+                )
+
+                results = []
+                for sub_idx in range(2):
+                    if config.attn_mask_type == NVTE_Mask_Type.NVTE_NO_MASK:
+                        k_unmasked, v_unmasked = k, v  # full kv used for unmasked
+                    else:
+                        k_unmasked, v_unmasked = helper.slice_kv(k, v, kv_seqlens_for_rank[sub_idx])
+
+                    q_seqlen = q_seqlen // (cp_size * 2)
+                    num_kv_chunks = kv_max_seqlen // kv_seqlens_for_rank[sub_idx]
+                    kv_seqlen = (kv_seqlen // (cp_size * 2)) * num_kv_chunks
+
+                    dq_local, dk_local, dv_local, dbias_local = FusedAttnBwdPrimitive.impl(
+                        q_split[sub_idx],
+                        k_unmasked,
+                        v_unmasked,
+                        bias,
+                        softmax_aux_split[sub_idx],
+                        rng_state,
+                        output_split[sub_idx],
+                        doutput_split[sub_idx],
+                        q_seqlen,
+                        kv_seqlen,
+                        q_seq_offsets,
+                        k_seq_offsets,
+                        config=config,
+                    )
+
+                    # pad dk/dv to be unsliced shape so we can reduce scatter over all ranks.
+                    if config.attn_mask_type != NVTE_Mask_Type.NVTE_NO_MASK:
+                        pad_length = kv_max_seqlen - kv_seqlens_for_rank[sub_idx]
+                        dk_local, dv_local = helper.pad_kv(dk_local, dv_local, pad_length)
+
+                    results.append((dq_local, dk_local, dv_local, dbias_local))
+
+                dq_local = jnp.concatenate((results[0][0], results[1][0]), axis=1)
+                dk_local_pad = results[0][1] + results[1][1]
+                dv_local_pad = results[0][2] + results[1][2]
+                return dq_local, dk_local_pad, dv_local_pad, results[1][3]
+
+            k_ag, v_ag = helper.all_gather_kv(k, v)
+
+            functions = [
+                partial(
+                    _cross_attn_bwd,
+                    idx,
+                    q,
+                    k_ag,
+                    v_ag,
+                    bias,
+                    softmax_aux,
+                    rng_state,
+                    output,
+                    doutput,
+                    q_seqlen,
+                    kv_seqlen,
+                )
+                for idx in range(cp_size)
+            ]
+
+            dq, dk_local, dv_local, dbias = lax.switch(cp_rank, functions)
+            dk, dv = helper.reduce_scatter_dkv(dk_local, dv_local)
+
+            return dq, dk, dv, dbias
+
+        return mesh, impl, out_shardings, arg_shardings
+
+
+register_primitive(FusedAttnCPWithAllGatherBwdPrimitive)
+
+
 def fused_attn_fwd(
```

Comment:
This is an important discussion point: Do we keep CP hidden behind a common function or add a separate flavor of `fused_attn_cp_allgather_fwd`. The though here was it makes sense to keep a common interface that naturally support CP. Other implementation strategies e.g. ring can be exposed via argument.